src/patches.rs

```diff
@@ -0,0 +1,304 @@
+        let mut tmp_cargo_file = NamedTempFile::new()
+            .map_err(|err| format!("Unable to create a temporary file: {}", err))?;
+        tmp_cargo_file
+            .write_all(patched_cargo_doc.to_string().as_bytes())
+            .map_err(|err| format!("Unable to write temporary file: {}", err))?;
+
+        copy_patches_to_remote(&build_path, &build_server, tmp_cargo_file, project_list);
+    }
+    Ok(())
+}
+
+fn locate_workspace_folder(mut crate_path: PathBuf) -> Result<PathBuf, String> {
+    let cargo = std::env::var("CARGO").unwrap_or("cargo".to_owned());
+    log::debug!("Checking workspace root of path {:?}", crate_path);
+    crate_path.push("Cargo.toml");
+    let output = Command::new(cargo)
+        .arg("locate-project")
+        .arg("--workspace")
+        .arg("--manifest-path")
+        .arg(crate_path.as_os_str().clone())
+        .output()
+        .map_err(|err| format!("Unable to execute cargo locate-project: {:?}", err))?;
+
+    if !output.status.success() {
+        return Err(format!("{:?}", output.status));
+    }
+
+    let output = String::from_utf8(output.stdout).map_err(|e| e.to_string())?;
+    let parsed = json::parse(&output).map_err(|e| e.to_string())?;
+    let root = parsed["root"].as_str().ok_or(String::from("no root"))?;
+    let mut result = PathBuf::from(root);
+
+    // Remove the trailing "/Cargo.toml"
+    result.pop();
+    Ok(result)
+}
+
+#[derive(Debug, Clone)]
+struct PatchProject {
+    pub name: OsString,
+    pub local_path: PathBuf,
+    pub remote_path: PathBuf,
+}
+
+impl PatchProject {
+    pub fn new(name: OsString, path: PathBuf, remote_path: PathBuf) -> Self {
+        PatchProject {
+            name,
+            local_path: path,
+            remote_path,
+        }
+    }
+}
+
+fn extract_patched_crates_and_adjust_toml<F: Fn(PathBuf) -> Result<PathBuf, String>>(
+    manifest_content: String,
+    locate_workspace: F,
+) -> Result<Option<(Document, Vec<PatchProject>)>, String> {
+    let mut manifest = manifest_content.parse::<Document>().map_err(|err| {
+        format!(
+            "Unable to parse Cargo.toml: {:?} content: {}",
+            err, manifest_content
+        )
+    })?;
+    let mut workspaces_to_copy: Vec<PatchProject> = Vec::new();
+
+    // A list of inline tables like
+    // { path = "/some/path" }
+    let patched_paths: Option<Vec<&mut InlineTable>> =
+        manifest["patch"].as_table_mut().map(|patch| {
+            patch
+                .iter_mut()
+                .filter_map(|(_, crate_table)| crate_table.as_table_mut())
+                .flat_map(|crate_table| {
+                    crate_table
+                        .iter_mut()
+                        .filter_map(|(_, patch_table)| patch_table.as_inline_table_mut())
+                })
+                .collect()
+        });
+
+    if patched_paths.is_none() {
+        log::debug!("No patches in project.");
+        return Ok(None);
+    }
```

Comment:
Maybe properly unwrap the `patched_paths` with `if let Some()`?

Reply:
I thought about this but did not want to add another level of identation here for readability.